src/cryptonote_core/service_node_quorum_cop.cpp

```diff
@@ -137,113 +154,159 @@ namespace service_nodes
-            time_t const now = time(nullptr);
-#if defined(LOKI_ENABLE_INTEGRATION_TEST_HOOKS)
-            time_t const min_lifetime = 0;
-#else
-            time_t const min_lifetime = 60 * 60 * 2;
-#endif
-            bool alive_for_min_time = (now - m_core.get_start_time()) >= min_lifetime;
-            if (!alive_for_min_time)
+            if (m_core.get_hard_fork_version(m_uptime_proof_height) < 9) continue;
+
+            const std::shared_ptr<const testing_quorum> quorum =
+                m_core.get_testing_quorum(quorum_type::uptime, m_uptime_proof_height);
+            if (!quorum)
             {
+              // TODO(loki): Fatal error
+              LOG_ERROR("Uptime quorum for height: " << m_uptime_proof_height << " was not cached in daemon!");
               continue;
             }
 
-            if (m_uptime_proof_height < start_voting_from_height)
-              m_uptime_proof_height = start_voting_from_height;
+            int index_in_group = find_index_in_quorum_group(quorum->validators, my_pubkey);
+            if (index_in_group <= -1) continue;
+
+            for (size_t node_index = 0; node_index < quorum->workers.size(); ++node_index)
+            {
+              const crypto::public_key &node_key = quorum->workers[node_index];
+
+              CRITICAL_REGION_LOCAL(m_lock);
+              bool vote_off_node = (m_uptime_proof_seen.find(node_key) == m_uptime_proof_seen.end());
+              if (!vote_off_node) continue;
+
+              quorum_vote_t vote = service_nodes::make_deregister_vote(hf_version,
+                                                                       service_nodes::make_deregister_type::uptime,
+                                                                       m_uptime_proof_height,
+                                                                       static_cast<uint16_t>(index_in_group),
+                                                                       node_index,
+                                                                       my_pubkey,
+                                                                       my_seckey);
+              cryptonote::vote_verification_context vvc = {};
+              if (!handle_vote(vote, vvc))
+                LOG_ERROR("Failed to add uptime deregister vote reason: " << print_vote_verification_context(vvc, &vote));
+            }
+          }
+        }
+        break;
+
+        case quorum_type::checkpointing:
+        {
+          if (hf_version < cryptonote::network_version_12_checkpointing)
+            break;
+
+          // NOTE: Cast deregister votes if service nodes have not participated
+          uint64_t const half_vote_lifetime                 = VOTE_LIFETIME / 2;
+          uint64_t const start_validating_checkpoint_height = start_voting_from_height;
+          uint64_t const end_validating_checkpoint_height =
```

Comment:
Was thinking this through and I think this is buggy. This would be the same bug as we had before with uptimes, we can't vote immediately until we collected the votes, since we don't accept votes while syncing- so we wait 2hrs before attempting anything in the uptime proof case.